### ISSUE TITLE
Uyuni Bugreport - Uyuni 2021.06 - Error while syncing EPEL 8 repository

### ISSUE THREAD

**Issue by jorx-dev** [uyuni-project/uyuni#4011]
# Additional Information

## Version of Uyuni Server and Proxy (if used)
`zypper info Uyuni-Server-release` 

LANG=C zypper info Uyuni-Server-release
Loading repository data...
Reading installed packages...

Information for package Uyuni-Server-release:
---------------------------------------------
Repository     : uyuni-server-stable
Name           : Uyuni-Server-release
Version        : 2021.06-132.3.uyuni1
Arch           : x86_64
Vendor         : obs://build.opensuse.org/systemsmanagement:Uyuni
Support Level  : Level 3
Installed Size : 1.4 KiB
Installed      : Yes (automatically)
Status         : up-to-date
Source package : Uyuni-Server-release-2021.06-132.3.uyuni1.src
Summary        : Uyuni Server
Description    : 
    Uyuni lets you efficiently manage physical, virtual,
    and cloud-based Linux systems. It provides automated and cost-effective
    configuration and software management, asset management, and system
    provisioning.

## Details about issue

Syncing EPEL 8 repositories fails. Some developers showed up at the users mailing list pointing that this issue is related to a postgresql constraint that limits what kind of a patch is ('final', 'stable', 'testing' or 'retracted'). This error is triggered when a kind of patch named "pending" is found in the repository. 

Uyuni users mailing list archive reference:

https://lists.opensuse.org/archives/list/users@lists.uyuni-project.org/thread/BL7JAO3EU3LTI5QANUO4P45J5YNGAXMI/

Full log:

# LANG=C spacewalk-repo-sync -c epel8-almalinux8-x86_64
08:46:02 ======================================
08:46:02 | Channel: epel8-almalinux8-x86_64
08:46:02 ======================================
08:46:02 Sync of channel started.
Retrieving repository 'epel8-almalinux8-x86_64' metadata
...............................[done]
Building repository 'epel8-almalinux8-x86_64' cache
....................................[done]
All repositories have been refreshed.
08:46:06 
08:46:06   Importing comps file
96d01da780a309a1803ea903a9f2a33b1c239a265847c6a86718acb9ef85f06e-comps-
Everything.x86_64.xml.
08:46:06   Renaming non-standard filename
96d01da780a309a1803ea903a9f2a33b1c239a265847c6a86718acb9ef85f06e-comps-
Everything.x86_64.xml to comps.x86_64.xml.
08:46:06 *** NOTE: Importing comps file for the channel 'epel8-
almalinux8-x86_64'. Previous comps will be discarded.
08:46:06 Repo URL:
https://dl.fedoraproject.org/pub/epel/8/Everything/x86_64/
08:46:06     Packages in repo:              7558
08:46:14     No new packages to sync.
08:46:15 
08:46:15   Patches in repo: 3017.
08:46:15 Add Patch FEDORA-EPEL-2021-3ec3998292
08:46:16 Add Patch FEDORA-EPEL-2021-7dc78a18ca
08:46:16 Add Patch FEDORA-EPEL-2021-d33da662e7
08:46:16 Add Patch FEDORA-EPEL-2021-87b303b50e
08:46:17 Add Patch FEDORA-EPEL-2021-7ed6a63b7b
08:46:17 Add Patch FEDORA-EPEL-2021-a5a0c05adf
08:46:17 Add Patch FEDORA-EPEL-2021-f77315a931
08:46:17 Add Patch FEDORA-EPEL-2021-333cf44938
08:46:17 Add Patch FEDORA-EPEL-2021-fdb6cf1e3a
08:46:17 Add Patch FEDORA-EPEL-2021-b3771b0cd1
08:46:18 Add Patch FEDORA-EPEL-2021-15c5b7660c
08:46:18 Add Patch FEDORA-EPEL-2021-384d2492fa
08:46:18 Add Patch FEDORA-EPEL-2021-247500b671
08:46:18 Add Patch FEDORA-EPEL-2021-d4c3bbff96
08:46:19 Add Patch FEDORA-EPEL-2021-a60d1222d2
08:46:19 Add Patch FEDORA-EPEL-2021-ec70035a87
08:46:19 Add Patch FEDORA-EPEL-2021-d48e8d2d55
08:46:19 Add Patch FEDORA-EPEL-2021-293195e59b
08:46:19 Add Patch FEDORA-EPEL-2021-41eeae14d9
08:46:19 Add Patch FEDORA-EPEL-2021-ec2a0a616a
08:46:19 Add Patch FEDORA-EPEL-2021-b8960f6143
08:46:19 Add Patch FEDORA-EPEL-2021-20a66b3776
08:46:20 Add Patch FEDORA-EPEL-2020-7c3fde7f74
08:46:20 Add Patch FEDORA-EPEL-2021-ac096c00b0
08:46:20 Add Patch FEDORA-EPEL-2021-aa67d44d29
08:46:20 Add Patch FEDORA-EPEL-2021-d3d8cfc98d
08:46:20     Syncing 25 new patch(es) to channel.
08:46:20 Unexpected error: <class 'psycopg2.errors.CheckViolation'>
08:46:20 Traceback (most recent call last):
  File "/usr/lib/python3.6/site-
packages/spacewalk/satellite_tools/reposync.py", line 600, in sync
    self.import_updates(plugin)
  File "/usr/lib/python3.6/site-
packages/spacewalk/satellite_tools/reposync.py", line 741, in
import_updates
    processed_updates_count = self.upload_updates(notices)
  File "/usr/lib/python3.6/site-
packages/spacewalk/satellite_tools/reposync.py", line 967, in
upload_updates
    importer.run()
  File "/usr/lib/python3.6/site-
packages/spacewalk/server/importlib/importLib.py", line 778, in run
    self.submit()
  File "/usr/lib/python3.6/site-
packages/spacewalk/server/importlib/errataImport.py", line 217, in
submit
    dml = self.backend.processErrata(self.batch)
  File "/usr/lib/python3.6/site-
packages/spacewalk/server/importlib/backend.py", line 1011, in
processErrata
    transactional=1)
  File "/usr/lib/python3.6/site-
packages/spacewalk/server/importlib/backend.py", line 2151, in
__processObjectCollection
    return self.__processObjectCollection__(objColl, parentTable,
childDict, **kwargs)
  File "/usr/lib/python3.6/site-
packages/spacewalk/server/importlib/backend.py", line 2322, in
__processObjectCollection__
    return self.__doDML(dml)
  File "/usr/lib/python3.6/site-
packages/spacewalk/server/importlib/backend.py", line 2427, in __doDML
    self.__doInsert(dml.insert, dml.tables)
  File "/usr/lib/python3.6/site-
packages/spacewalk/server/importlib/backend.py", line 2434, in
__doInsert
    self.__doInsertTable(tname, dict)
  File "/usr/lib/python3.6/site-
packages/spacewalk/server/importlib/backend.py", line 2449, in
__doInsertTable
    insertObj.query(hash)
  File "/usr/lib/python3.6/site-
packages/spacewalk/server/importlib/backendLib.py", line 453, in query
    statement.execute_values(self._buildQuery(None), value_list,
fetch=False, page_size=10_000)
  File "/usr/lib/python3.6/site-
packages/spacewalk/server/rhnSQL/sql_base.py", line 166, in
execute_values
    return self._execute_wrapper(self._execute_values, sql, argslist,
template, page_size, fetch)
  File "/usr/lib/python3.6/site-
packages/spacewalk/server/rhnSQL/driver_postgresql.py", line 299, in
_execute_wrapper
    retval = function(*p, **kw)
  File "/usr/lib/python3.6/site-
packages/spacewalk/server/rhnSQL/driver_postgresql.py", line 353, in
_execute_values
    results = psycopg2.extras.execute_values(self._real_cursor, sql,
argslist, template=template, page_size=page_size, fetch=fetch)
  File "/usr/lib64/python3.6/site-packages/psycopg2/extras.py", line
1289, in execute_values
    cur.execute(b''.join(parts))
psycopg2.errors.CheckViolation: new row for relation "rhnerrata"
violates check constraint "rhn_errata_adv_status_ck"
DETAIL:  Failing row contains (8816, FEDORA-EPEL-2021-a5a0c05adf,
Product Enhancement Advisory, FEDORA-EPEL-2021-a5a0c05adf, 200,
pending, Fedora EPEL 8, Initial package, python-jdcal-1.4.1-1.el8,  , 
, 2021-07-19 16:46:28-02, 2021-07-19 16:46:28-02, null, null, 1, null,
updates@fedoraproject.org, 2021-07-20 04:46:02.984641-02, 2021-07-20
04:46:02.984641-02, 2021-07-20 04:46:02.984641-02, null).

08:46:20 Total time: 0:00:17

**Comment by digdilem** [comment]
Copied from mailing list reply;

I'm NOT seeing the same issue when syncing Epel for Centos-8 or Rocky-8 (same repo included in different channels)

It's behaving normally here.  Will add a note to the issue as well, in case it helps the devs.
 

```
2021/03/21 14:00:13 +01:00 Repo URL: http://www.mirrorservice.org/sites/dl.fedoraproject.org/pub/epel/8/Everything/x86_64/
2021/03/21 14:00:13 +01:00     Packages in repo:              7121
2021/03/21 14:00:27 +01:00     Packages already synced:       6941
2021/03/21 14:00:27 +01:00     Packages to sync:               180
2021/03/21 14:00:27 +01:00     New packages to download:       180
2021/03/21 14:00:27 +01:00   Downloading packages:
2021/03/21 14:00:28 +01:00     1/180 : asciinema-2.0.2-6.el8.noarch.rpm
2021/03/21 14:00:28 +01:00     2/180 : distribution-gpg-keys-1.51-1.el8.noarch.rpm

...

2021/03/21 14:00:41 +01:00     180/180 : zabbix40-web-4.0.29-1.el8.noarch.rpm
2021/03/21 14:00:41 +01:00 Importing packages started.
2021/03/21 14:00:41 +01:00
2021/03/21 14:00:41 +01:00   Importing packages to DB:
2021/03/21 14:00:45 +01:00   Package batch #3 of 9 completed...

...

2021/03/21 14:01:11 +01:00   Package batch #8 of 9 completed...
2021/03/21 14:01:11 +01:00 Importing packages finished.
2021/03/21 14:01:11 +01:00

...

2021/07/21 08:38:27 +01:00   Linking packages to the channel.
2021/07/21 08:38:28 +01:00     34 packages linked
2021/07/21 08:38:29 +01:00
2021/07/21 08:38:29 +01:00   Patches in repo: 3018.
2021/07/21 08:38:30 +01:00 Add Patch FEDORA-EPEL-2021-762ccf9ade
2021/07/21 08:38:31 +01:00 Add Patch FEDORA-EPEL-2021-a60d1222d2
2021/07/21 08:38:31 +01:00 Add Patch FEDORA-EPEL-2021-aa67d44d29
2021/07/21 08:38:34 +01:00 Add Patch FEDORA-EPEL-2021-ec2a0a616a
2021/07/21 08:38:35 +01:00 Add Patch FEDORA-EPEL-2020-7c3fde7f74
2021/07/21 08:38:36 +01:00 Add Patch FEDORA-EPEL-2021-7ed6a63b7b
2021/07/21 08:38:36 +01:00 Add Patch FEDORA-EPEL-2021-41eeae14d9
2021/07/21 08:38:36 +01:00 Add Patch FEDORA-EPEL-2021-9ba6d5793a
2021/07/21 08:38:37 +01:00 Add Patch FEDORA-EPEL-2021-a5a0c05adf
2021/07/21 08:38:37 +01:00 Add Patch FEDORA-EPEL-2021-ec70035a87
2021/07/21 08:38:37 +01:00     Syncing 9 new patch(es) to channel.
2021/07/21 08:38:41 +01:00
2021/07/21 08:38:41 +01:00   Importing comps file bbe4d6c0eadd96f6a033794a25f89a2df43e5a02007f8fe748ef7634099bcbbc-comps-BaseOS.x86_64.xml.
2021/07/21 08:38:41 +01:00   Renaming non-standard filename bbe4d6c0eadd96f6a033794a25f89a2df43e5a02007f8fe748ef7634099bcbbc-comps-BaseOS.x86_64.xml to comps.x86_64.xml.
```

**Comment by nkrinner** [comment]
@cbbayburt  FYI

**Comment by digdilem** [comment]
Update: I AM now seeing this issue in EPEL 8....

Can't explain why it took so long to affect me, but it started here about a week ago. Even after setting up a new repo it persists, breaking on the same patch each time:

```
2021/09/08 11:15:40 +01:00 Add Patch FEDORA-EPEL-2018-81bb24fe00
2021/09/08 11:15:40 +01:00 Add Patch FEDORA-EPEL-2018-6f7d6eeeed
2021/09/08 11:15:40 +01:00     Syncing 4684 new patch(es) to channel.
2021/09/08 11:15:58 +01:00 Unexpected error: <class 'psycopg2.errors.CheckViolation'>
2021/09/08 11:15:58 +01:00 Traceback (most recent call last):
  File "/usr/lib/python3.6/site-packages/spacewalk/satellite_tools/reposync.py", line 600, in sync
    self.import_updates(plugin)
  File "/usr/lib/python3.6/site-packages/spacewalk/satellite_tools/reposync.py", line 741, in import_updates
    processed_updates_count = self.upload_updates(notices)
  File "/usr/lib/python3.6/site-packages/spacewalk/satellite_tools/reposync.py", line 967, in upload_updates
    importer.run()
  File "/usr/lib/python3.6/site-packages/spacewalk/server/importlib/importLib.py", line 778, in run
    self.submit()
  File "/usr/lib/python3.6/site-packages/spacewalk/server/importlib/errataImport.py", line 217, in submit
    dml = self.backend.processErrata(self.batch)
  File "/usr/lib/python3.6/site-packages/spacewalk/server/importlib/backend.py", line 1011, in processErrata
    transactional=1)
  File "/usr/lib/python3.6/site-packages/spacewalk/server/importlib/backend.py", line 2151, in __processObjectCollection
    return self.__processObjectCollection__(objColl, parentTable, childDict, **kwargs)
  File "/usr/lib/python3.6/site-packages/spacewalk/server/importlib/backend.py", line 2322, in __processObjectCollection__
    return self.__doDML(dml)
  File "/usr/lib/python3.6/site-packages/spacewalk/server/importlib/backend.py", line 2427, in __doDML
    self.__doInsert(dml.insert, dml.tables)
  File "/usr/lib/python3.6/site-packages/spacewalk/server/importlib/backend.py", line 2434, in __doInsert
    self.__doInsertTable(tname, dict)
  File "/usr/lib/python3.6/site-packages/spacewalk/server/importlib/backend.py", line 2449, in __doInsertTable
    insertObj.query(hash)
  File "/usr/lib/python3.6/site-packages/spacewalk/server/importlib/backendLib.py", line 453, in query
    statement.execute_values(self._buildQuery(None), value_list, fetch=False, page_size=10_000)
  File "/usr/lib/python3.6/site-packages/spacewalk/server/rhnSQL/sql_base.py", line 166, in execute_values
    return self._execute_wrapper(self._execute_values, sql, argslist, template, page_size, fetch)
  File "/usr/lib/python3.6/site-packages/spacewalk/server/rhnSQL/driver_postgresql.py", line 299, in _execute_wrapper
    retval = function(*p, **kw)
  File "/usr/lib/python3.6/site-packages/spacewalk/server/rhnSQL/driver_postgresql.py", line 353, in _execute_values
    results = psycopg2.extras.execute_values(self._real_cursor, sql, argslist, template=template, page_size=page_size, fetch=fetch)
  File "/usr/lib64/python3.6/site-packages/psycopg2/extras.py", line 1289, in execute_values
    cur.execute(b''.join(parts))
psycopg2.errors.CheckViolation: new row for relation "rhnerrata" violates check constraint "rhn_errata_adv_status_ck"
DETAIL:  Failing row contains (11845, FEDORA-EPEL-2021-aa19cf2506, Product Enhancement Advisory, FEDORA-EPEL-2021-aa19cf2506, 200, Fedora EPEL 7, Rebuild for Knot DNS 3.1, knot-resolver-5.4.1-2.el7,  ,  , 2021-09-06 03:52:14-01, 2021-09-06 03:52:14-01, null, null, 1, null, updates@fedoraproject.org, 2021-09-08 09:14:29.766441-01, 2021-09-08 09:14:29.766441-01, 2021-09-08 09:14:29.766441-01, 1, pending).
```

**Comment by mcalmer** [comment]
The status "pending" for an errata is not so usual. We have a fixed list of valid status items and "pending" was not part of it.
I think tomorrow you can mirror as this errata change its status to "final".
The fix is already made, but not yet released. I think the next uyuni release should have this fixed and allow also "pending".

While I am not sure if you want to use an errata which is "pending".

**Comment by cbbayburt** [comment]
Fixed via #4134. The fix will be available in the next release.
